compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,320 @@
+            continue;
+        }
+
+        let mut finder = MutatedLocalFinder { local: discr, mutated: false };
+        'preds: for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let terminator = data.terminator();
+            let TerminatorKind::Goto { .. } = terminator.kind else {
+                continue
+            };
+
+            for (statement_index, statement) in data.statements.iter().enumerate().rev() {
+                match statement.kind {
+                    StatementKind::SetDiscriminant { box place, variant_index: variant }
+                    | StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(_, variant, ..), ..),
+                    )) if place.local == discr => {
+                        if finder.mutated {
+                            debug!(
+                                "The discriminant: {discr:?} was mutated in predecessor: {pred:?}"
+                            );
+                            // We can't optimize this predecessor, so try the next one.
+                            finder.mutated = false;
+
+                            continue 'preds;
+                        }
+
+                        let discr_ty = body.local_decls[discr].ty;
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            debug!(
+                                "{pred:?}: {place:?} = {discr_ty:?}::{variant:?}; goto -> {block:?}",
+                            );
+
+                            let target = targets.target_for_value(discr.val);
+                            static_switches
+                                .entry(block)
+                                .and_modify(|static_switches: &mut &mut [StaticSwitch]| {
+                                    if static_switches.iter_mut().all(|switch| {
+                                        if switch.pred == pred {
+                                            switch.target = target;
+                                            false
+                                        } else {
+                                            true
+                                        }
+                                    }) {
+                                        *static_switches =
+                                            tcx.arena.alloc_from_iter(
+                                                static_switches.iter().copied().chain([
+                                                    StaticSwitch { pred, target, exclude },
+                                                ]),
+                                            );
+                                    }
+                                })
+                                .or_insert_with(|| {
+                                    tcx.arena.alloc([StaticSwitch { pred, target, exclude }])
+                                });
```

Comment:
I don't understand how the worklist is stored and why.
Could you add an explanation?
Why isn't a vector of edges `(pred, block, target)`, meaning "the pred->block edge becomes pred->target" enough?